File: identity/derive_test.go
package identity

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"
)

func TestDeriveResponseKeys(t *testing.T) {
	// Test vectors with sequential byte values for reproducibility
	exportedSecret := make([]byte, 32)
	for i := range exportedSecret {
		exportedSecret[i] = byte(i)
	}

	requestEnc := make([]byte, 32)
	for i := range requestEnc {
		requestEnc[i] = byte(i + 32)
	}

	responseNonce := make([]byte, ResponseNonceLength)
	for i := range responseNonce {
		responseNonce[i] = byte(i + 64)
	}

	km, err := DeriveResponseKeys(exportedSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("DeriveResponseKeys failed: %v", err)
	}

	// Verify key length
	if len(km.Key) != AES256KeyLength {
		t.Errorf("Expected key length %d, got %d", AES256KeyLength, len(km.Key))
	}

	// Verify nonce base length
	if len(km.NonceBase) != AESGCMNonceLength {
		t.Errorf("Expected nonce base length %d, got %d", AESGCMNonceLength, len(km.NonceBase))
	}

	// Verify determinism: same inputs should produce same outputs
	km2, err := DeriveResponseKeys(exportedSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("Second DeriveResponseKeys failed: %v", err)
	}
	if !bytes.Equal(km.Key, km2.Key) {
		t.Error("Key derivation is not deterministic")
	}
	if !bytes.Equal(km.NonceBase, km2.NonceBase) {
		t.Error("Nonce base derivation is not deterministic")
	}

	// Verify different inputs produce different outputs
	differentNonce := make([]byte, ResponseNonceLength)
	copy(differentNonce, responseNonce)
	differentNonce[0] ^= 0xFF

	km3, err := DeriveResponseKeys(exportedSecret, requestEnc, differentNonce)
	if err != nil {
		t.Fatalf("Third DeriveResponseKeys failed: %v", err)
	}
	if bytes.Equal(km.Key, km3.Key) {
		t.Error("Different nonces should produce different keys")
	}
	if bytes.Equal(km.NonceBase, km3.NonceBase) {
		t.Error("Different nonces should produce different nonce bases")
	}

	// Verify different request enc produces different keys
	differentEnc := make([]byte, 32)
	copy(differentEnc, requestEnc)
	differentEnc[0] ^= 0xFF

	km4, err := DeriveResponseKeys(exportedSecret, differentEnc, responseNonce)
	if err != nil {
		t.Fatalf("Fourth DeriveResponseKeys failed: %v", err)
	}
	if bytes.Equal(km.Key, km4.Key) {
		t.Error("Different request enc should produce different keys")
	}

	// Verify different exported secret produces different keys
	differentSecret := make([]byte, 32)
	copy(differentSecret, exportedSecret)
	differentSecret[0] ^= 0xFF

	km5, err := DeriveResponseKeys(differentSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("Fifth DeriveResponseKeys failed: %v", err)
	}
	if bytes.Equal(km.Key, km5.Key) {
		t.Error("Different exported secret should produce different keys")
	}
}

func TestNonceComputation(t *testing.T) {
	// Create a ResponseAEAD with a known nonce base
	km := &ResponseKeyMaterial{
		Key:       make([]byte, 32), // Valid AES-256 key
		NonceBase: make([]byte, 12),
	}
	for i := range km.NonceBase {
		km.NonceBase[i] = 0xFF
	}

	aead, err := km.NewResponseAEAD()
	if err != nil {
		t.Fatalf("NewResponseAEAD failed: %v", err)
	}

	// Sequence 0 should return the base nonce
	nonce0 := aead.NonceForSeq(0)
	if !bytes.Equal(nonce0, km.NonceBase) {
		t.Error("Sequence 0 should return base nonce")
	}

	// Sequence 1 should differ in the last byte
	nonce1 := aead.NonceForSeq(1)
	if nonce1[11] != 0xFE { // 0xFF XOR 0x01
		t.Errorf("Expected last byte to be 0xFE, got 0x%02X", nonce1[11])
	}

	// Verify first 4 bytes are unchanged for small sequence numbers
	for i := range 4 {
		if nonce1[i] != 0xFF {
			t.Errorf("Byte %d should be 0xFF for seq=1, got 0x%02X", i, nonce1[i])
		}
	}

	// Verify all nonces are unique for first 1000 sequences
	seen := make(map[string]bool)
	for i := range uint64(1000) {
		nonce := aead.NonceForSeq(i)
		key := string(nonce)
		if seen[key] {
			t.Errorf("Duplicate nonce at sequence %d", i)
		}
		seen[key] = true
	}
}

func TestNonceLargeSequence(t *testing.T) {
	km := &ResponseKeyMaterial{
		Key:       make([]byte, 32), // Valid AES-256 key
		NonceBase: make([]byte, 12),
	}

	aead, err := km.NewResponseAEAD()
	if err != nil {
		t.Fatalf("NewResponseAEAD failed: %v", err)
	}

	// Test with a large sequence number that uses multiple bytes
	seq := uint64(0x0102030405060708)
	nonce := aead.NonceForSeq(seq)

	// Verify XOR was applied correctly to last 8 bytes
	expected := []byte{0, 0, 0, 0, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08}
	if !bytes.Equal(nonce, expected) {
		t.Errorf("Large sequence nonce incorrect.\nExpected: %x\nGot:      %x", expected, nonce)
	}
}

func TestDeriveResponseKeysValidation(t *testing.T) {
	validSecret := make([]byte, 32)
	validEnc := make([]byte, 32)
	validNonce := make([]byte, ResponseNonceLength)

	// Test invalid secret length
	_, err := DeriveResponseKeys(make([]byte, 16), validEnc, validNonce)
	if err == nil {
		t.Error("Expected error for invalid secret length")
	}

	// Test invalid enc length
	_, err = DeriveResponseKeys(validSecret, make([]byte, 16), validNonce)
	if err == nil {
		t.Error("Expected error for invalid enc length")
	}

	// Test invalid nonce length
	_, err = DeriveResponseKeys(validSecret, validEnc, make([]byte, 8))
	if err == nil {
		t.Error("Expected error for invalid nonce length")
	}

	// Test empty inputs
	_, err = DeriveResponseKeys(nil, validEnc, validNonce)
	if err == nil {
		t.Error("Expected error for nil secret")
	}

	_, err = DeriveResponseKeys(validSecret, nil, validNonce)
	if err == nil {
		t.Error("Expected error for nil enc")
	}

	_, err = DeriveResponseKeys(validSecret, validEnc, nil)
	if err == nil {
		t.Error("Expected error for nil nonce")
	}
}

func TestNewResponseAEAD(t *testing.T) {
	exportedSecret := make([]byte, 32)
	requestEnc := make([]byte, 32)
	responseNonce := make([]byte, ResponseNonceLength)

	km, err := DeriveResponseKeys(exportedSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("DeriveResponseKeys failed: %v", err)
	}

	aead, err := km.NewResponseAEAD()
	if err != nil {
		t.Fatalf("NewResponseAEAD failed: %v", err)
	}

	// Verify AEAD has correct properties
	if aead.NonceSize() != AESGCMNonceLength {
		t.Errorf("Expected nonce size %d, got %d", AESGCMNonceLength, aead.NonceSize())
	}

	// Verify overhead is 16 bytes (GCM tag)
	if aead.Overhead() != 16 {
		t.Errorf("Expected overhead 16, got %d", aead.Overhead())
	}
}

func TestEncryptDecryptRoundTrip(t *testing.T) {
	exportedSecret := make([]byte, 32)
	requestEnc := make([]byte, 32)
	responseNonce := make([]byte, ResponseNonceLength)

	// Fill with non-zero values for a more realistic test
	for i := range exportedSecret {
		exportedSecret[i] = byte(i * 3)
	}
	for i := range requestEnc {
		requestEnc[i] = byte(i * 5)
	}
	for i := range responseNonce {
		responseNonce[i] = byte(i * 7)
	}

	// Use same derivation for both "server" and "client"
	serverKM, err := DeriveResponseKeys(exportedSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("Server DeriveResponseKeys failed: %v", err)
	}

	clientKM, err := DeriveResponseKeys(exportedSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("Client DeriveResponseKeys failed: %v", err)
	}

	serverAEAD, err := serverKM.NewResponseAEAD()
	if err != nil {
		t.Fatalf("Server NewResponseAEAD failed: %v", err)
	}

	clientAEAD, err := clientKM.NewResponseAEAD()
	if err != nil {
		t.Fatalf("Client NewResponseAEAD failed: %v", err)
	}

	// Server encrypts (sequence auto-increments internally)
	plaintext := []byte("Hello, this is a secret response!")
	ciphertext := serverAEAD.Seal(plaintext, nil)

	// Client decrypts (sequence auto-increments internally)
	decrypted, err := clientAEAD.Open(ciphertext, nil)
	if err != nil {
		t.Fatalf("Decryption failed: %v", err)
	}

	if !bytes.Equal(plaintext, decrypted) {
		t.Errorf("Decrypted text doesn't match original.\nExpected: %s\nGot: %s", plaintext, decrypted)
	}
}

func TestMultipleChunksRoundTrip(t *testing.T) {
	exportedSecret := make([]byte, 32)
	requestEnc := make([]byte, 32)
	responseNonce := make([]byte, ResponseNonceLength)

	km, err := DeriveResponseKeys(exportedSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("DeriveResponseKeys failed: %v", err)
	}

	// Create separate AEAD instances for server and client
	// (both derived from the same key material)
	serverAEAD, err := km.NewResponseAEAD()
	if err != nil {
		t.Fatalf("Server NewResponseAEAD failed: %v", err)
	}

	clientAEAD, err := km.NewResponseAEAD()
	if err != nil {
		t.Fatalf("Client NewResponseAEAD failed: %v", err)
	}

	// Encrypt multiple chunks (sequence auto-increments: 0, 1, 2)
	chunks := [][]byte{
		[]byte("First chunk of data"),
		[]byte("Second chunk"),
		[]byte("Third and final chunk with more data"),
	}

	var ciphertexts [][]byte
	for _, chunk := range chunks {
		ct := serverAEAD.Seal(chunk, nil)
		ciphertexts = append(ciphertexts, ct)
	}

	// Verify server sequence is now 3
	if serverAEAD.Seq() != 3 {
		t.Errorf("Expected server sequence 3, got %d", serverAEAD.Seq())
	}

	// Decrypt all chunks (sequence auto-increments: 0, 1, 2)
	for i, ct := range ciphertexts {
		decrypted, err := clientAEAD.Open(ct, nil)
		if err != nil {
			t.Fatalf("Failed to decrypt chunk %d: %v", i, err)
		}
		if !bytes.Equal(decrypted, chunks[i]) {
			t.Errorf("Chunk %d mismatch.\nExpected: %s\nGot: %s", i, chunks[i], decrypted)
		}
	}

	// Verify client sequence is now 3
	if clientAEAD.Seq() != 3 {
		t.Errorf("Expected client sequence 3, got %d", clientAEAD.Seq())
	}
}

func TestWrongSequenceNumberFails(t *testing.T) {
	exportedSecret := make([]byte, 32)
	requestEnc := make([]byte, 32)
	responseNonce := make([]byte, ResponseNonceLength)

	km, err := DeriveResponseKeys(exportedSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("DeriveResponseKeys failed: %v", err)
	}

	aead, err := km.NewResponseAEAD()
	if err != nil {
		t.Fatalf("NewResponseAEAD failed: %v", err)
	}

	// Encrypt with sequence 0 (auto-incremented)
	plaintext := []byte("Secret message")
	ciphertext := aead.Seal(plaintext, nil)

	// Try to decrypt with wrong sequence number using OpenWithSeq
	_, err = aead.OpenWithSeq(1, ciphertext, nil)
	if err == nil {
		t.Error("Decryption should fail with wrong sequence number")
	}

	// Verify correct sequence works
	_, err = aead.OpenWithSeq(0, ciphertext, nil)
	if err != nil {
		t.Errorf("Decryption should succeed with correct sequence number: %v", err)
	}
}

func TestDifferentKeysCannotDecrypt(t *testing.T) {
	// Server's key material
	serverSecret := make([]byte, 32)
	for i := range serverSecret {
		serverSecret[i] = byte(i)
	}

	requestEnc := make([]byte, 32)
	responseNonce := make([]byte, ResponseNonceLength)

	serverKM, err := DeriveResponseKeys(serverSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("Server DeriveResponseKeys failed: %v", err)
	}

	serverAEAD, err := serverKM.NewResponseAEAD()
	if err != nil {
		t.Fatalf("Server NewResponseAEAD failed: %v", err)
	}

	// Attacker's key material (different secret)
	attackerSecret := make([]byte, 32)
	for i := range attackerSecret {
		attackerSecret[i] = byte(i + 100) // Different values
	}

	attackerKM, err := DeriveResponseKeys(attackerSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("Attacker DeriveResponseKeys failed: %v", err)
	}

	attackerAEAD, err := attackerKM.NewResponseAEAD()
	if err != nil {
		t.Fatalf("Attacker NewResponseAEAD failed: %v", err)
	}

	// Server encrypts (sequence auto-increments)
	plaintext := []byte("Secret response only for legitimate client")
	ciphertext := serverAEAD.Seal(plaintext, nil)

	// Attacker tries to decrypt with their keys (using sequence 0)
	_, err = attackerAEAD.Open(ciphertext, nil)
	if err == nil {
		t.Error("Attacker should not be able to decrypt with different keys")
	}
}

func TestDeriveInteropVector(t *testing.T) {
	vectorJSON, err := os.ReadFile("../test-vectors/derive.json")
	if err != nil {
		t.Fatalf("Failed to read vector file: %v", err)
	}

	var vector struct {
		ExportedSecret string `json:"exportedSecret"`
		RequestEnc     string `json:"requestEnc"`
		ResponseNonce  string `json:"responseNonce"`
		DerivedKey     string `json:"derivedKey"`
		DerivedNonce   string `json:"derivedNonceBase"`
	}
	if err := json.Unmarshal(vectorJSON, &vector); err != nil {
		t.Fatalf("Failed to parse vector JSON: %v", err)
	}

	exportedSecret, _ := hex.DecodeString(vector.ExportedSecret)
	requestEnc, _ := hex.DecodeString(vector.RequestEnc)
	responseNonce, _ := hex.DecodeString(vector.ResponseNonce)
	expectedKey, _ := hex.DecodeString(vector.DerivedKey)
	expectedNonce, _ := hex.DecodeString(vector.DerivedNonce)

	km, err := DeriveResponseKeys(exportedSecret, requestEnc, responseNonce)
	if err != nil {
		t.Fatalf("DeriveResponseKeys failed: %v", err)
	}

	if !bytes.Equal(km.Key, expectedKey) {
		t.Errorf("Key mismatch.\nExpected: %s\nGot:      %s", vector.DerivedKey, hex.EncodeToString(km.Key))
	}
	if !bytes.Equal(km.NonceBase, expectedNonce) {
		t.Errorf("NonceBase mismatch.\nExpected: %s\nGot:      %s", vector.DerivedNonce, hex.EncodeToString(km.NonceBase))
	}
}

File: js/src/test/derive.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert';
import {
  deriveResponseKeys,
  computeNonce,
  encryptChunk,
  decryptChunk,
  hexToBytes,
  bytesToHex,
  EXPORT_LENGTH,
  RESPONSE_NONCE_LENGTH,
  AES_GCM_NONCE_LENGTH,
  REQUEST_ENC_LENGTH,
} from '../derive.js';

describe('deriveResponseKeys', () => {
  it('should derive deterministic keys', async () => {
    const exportedSecret = new Uint8Array(32).fill(1);
    const requestEnc = new Uint8Array(32).fill(2);
    const responseNonce = new Uint8Array(32).fill(3);

    const km1 = await deriveResponseKeys(exportedSecret, requestEnc, responseNonce);
    const km2 = await deriveResponseKeys(exportedSecret, requestEnc, responseNonce);

    assert.deepStrictEqual(km1.key, km2.key, 'Keys should be identical');
    assert.deepStrictEqual(km1.nonceBase, km2.nonceBase, 'Nonce bases should be identical');
  });

  it('should produce different keys for different inputs', async () => {
    const exportedSecret = new Uint8Array(32).fill(1);
    const requestEnc = new Uint8Array(32).fill(2);
    const responseNonce1 = new Uint8Array(32).fill(3);
    const responseNonce2 = new Uint8Array(32).fill(4);

    const km1 = await deriveResponseKeys(exportedSecret, requestEnc, responseNonce1);
    const km2 = await deriveResponseKeys(exportedSecret, requestEnc, responseNonce2);

    assert.notDeepStrictEqual(km1.key, km2.key, 'Keys should differ for different nonces');
  });

  it('should reject invalid exported secret length', async () => {
    const shortSecret = new Uint8Array(16);
    const requestEnc = new Uint8Array(32);
    const responseNonce = new Uint8Array(32);

    await assert.rejects(
      async () => deriveResponseKeys(shortSecret, requestEnc, responseNonce),
      /exported secret must be 32 bytes/
    );
  });

  it('should reject invalid request enc length', async () => {
    const exportedSecret = new Uint8Array(32);
    const shortEnc = new Uint8Array(16);
    const responseNonce = new Uint8Array(32);

    await assert.rejects(
      async () => deriveResponseKeys(exportedSecret, shortEnc, responseNonce),
      /request enc must be 32 bytes/
    );
  });

  it('should reject invalid response nonce length', async () => {
    const exportedSecret = new Uint8Array(32);
    const requestEnc = new Uint8Array(32);
    const shortNonce = new Uint8Array(12); // Wrong - should be 32

    await assert.rejects(
      async () => deriveResponseKeys(exportedSecret, requestEnc, shortNonce),
      /response nonce must be 32 bytes/
    );
  });
});

describe('computeNonce', () => {
  it('should return base nonce for sequence 0', () => {
    const nonceBase = new Uint8Array(12).fill(0xFF);
    const nonce = computeNonce(nonceBase, 0);
    assert.deepStrictEqual(nonce, nonceBase, 'Nonce should equal base for seq 0');
  });

  it('should XOR correctly for sequence 1', () => {
    const nonceBase = new Uint8Array(12).fill(0xFF);
    const nonce = computeNonce(nonceBase, 1);
    assert.strictEqual(nonce[11], 0xFE, 'Last byte should be 0xFF XOR 0x01 = 0xFE');
  });

  it('should produce unique nonces for different sequences', () => {
    const nonceBase = new Uint8Array(12).fill(0);
    const seen = new Set<string>();

    for (let i = 0; i < 1000; i++) {
      const nonce = computeNonce(nonceBase, i);
      const key = bytesToHex(nonce);
      assert(!seen.has(key), `Nonce for seq ${i} should be unique`);
      seen.add(key);
    }
  });

  it('should handle large sequence numbers', () => {
    const nonceBase = new Uint8Array(12).fill(0);
    // Test doesn't throw for large numbers
    const nonce = computeNonce(nonceBase, 0xFFFFFFFF);
    assert.strictEqual(nonce.length, 12);
  });
});

describe('encrypt/decrypt round trip', () => {
  it('should round-trip successfully', async () => {
    const exportedSecret = new Uint8Array(32);
    crypto.getRandomValues(exportedSecret);
    const requestEnc = new Uint8Array(32);
    crypto.getRandomValues(requestEnc);
    const responseNonce = new Uint8Array(32);
    crypto.getRandomValues(responseNonce);

    const km = await deriveResponseKeys(exportedSecret, requestEnc, responseNonce);

    const plaintext = new TextEncoder().encode('Hello, World!');
    const ciphertext = await encryptChunk(km, 0, plaintext);
    const decrypted = await decryptChunk(km, 0, ciphertext);

    assert.deepStrictEqual(decrypted, plaintext, 'Decrypted should match original');
  });

  it('should fail with wrong sequence number', async () => {
    const exportedSecret = new Uint8Array(32);
    crypto.getRandomValues(exportedSecret);
    const requestEnc = new Uint8Array(32);
    const responseNonce = new Uint8Array(32);

    const km = await deriveResponseKeys(exportedSecret, requestEnc, responseNonce);

    const plaintext = new TextEncoder().encode('Hello, World!');
    const ciphertext = await encryptChunk(km, 0, plaintext);

    // Try to decrypt with wrong sequence
    await assert.rejects(
      async () => decryptChunk(km, 1, ciphertext),
      /error/i // AES-GCM decryption failure
    );
  });

  it('should encrypt multiple chunks with different nonces', async () => {
    const exportedSecret = new Uint8Array(32);
    crypto.getRandomValues(exportedSecret);
    const requestEnc = new Uint8Array(32);
    crypto.getRandomValues(requestEnc);
    const responseNonce = new Uint8Array(32);
    crypto.getRandomValues(responseNonce);

    const km = await deriveResponseKeys(exportedSecret, requestEnc, responseNonce);

    const chunks = ['chunk1', 'chunk2', 'chunk3'].map((s) =>
      new TextEncoder().encode(s)
    );

    // Encrypt all chunks
    const encrypted = await Promise.all(
      chunks.map((chunk, i) => encryptChunk(km, i, chunk))
    );

    // Decrypt all chunks
    const decrypted = await Promise.all(
      encrypted.map((ct, i) => decryptChunk(km, i, ct))
    );

    for (let i = 0; i < chunks.length; i++) {
      assert.deepStrictEqual(decrypted[i], chunks[i], `Chunk ${i} should match`);
    }
  });
});

describe('hex utilities', () => {
  it('should convert hex to bytes correctly', () => {
    const hex = 'deadbeef';
    const bytes = hexToBytes(hex);
    assert.deepStrictEqual(bytes, new Uint8Array([0xde, 0xad, 0xbe, 0xef]));
  });

  it('should convert bytes to hex correctly', () => {
    const bytes = new Uint8Array([0xde, 0xad, 0xbe, 0xef]);
    const hex = bytesToHex(bytes);
    assert.strictEqual(hex, 'deadbeef');
  });

  it('should round-trip hex conversion', () => {
    const original = new Uint8Array(32);
    crypto.getRandomValues(original);
    const hex = bytesToHex(original);
    const restored = hexToBytes(hex);
    assert.deepStrictEqual(restored, original);
  });

  it('should reject odd-length hex strings', () => {
    assert.throws(() => hexToBytes('abc'), /even length/);
  });
});

describe('constants', () => {
  it('should have correct constant values', () => {
    assert.strictEqual(EXPORT_LENGTH, 32);
    assert.strictEqual(RESPONSE_NONCE_LENGTH, 32);
    assert.strictEqual(AES_GCM_NONCE_LENGTH, 12);
    assert.strictEqual(REQUEST_ENC_LENGTH, 32);
  });
});

describe('Go interoperability', () => {
  it('should derive same keys as Go implementation', async () => {
    // Test vectors from Go tests: exportedSecret[i] = i, requestEnc[i] = i+32, responseNonce[i] = i+64
    const exportedSecret = new Uint8Array(32);
    for (let i = 0; i < 32; i++) exportedSecret[i] = i;

    const requestEnc = new Uint8Array(32);
    for (let i = 0; i < 32; i++) requestEnc[i] = i + 32;

    const responseNonce = new Uint8Array(32);
    for (let i = 0; i < 32; i++) responseNonce[i] = i + 64;

    const km = await deriveResponseKeys(exportedSecret, requestEnc, responseNonce);

    // Expected values from Go implementation
    const expectedKey = hexToBytes('40ec528847cd4e928449f2ed1a70a7d1e8ee317d5e900424fc1dd5b0475b97f7');
    const expectedNonceBase = hexToBytes('f8b0ce9466f27aa6243c65f9');

    assert.deepStrictEqual(
      km.key,
      expectedKey,
      `Key mismatch.\nExpected: ${bytesToHex(expectedKey)}\nGot: ${bytesToHex(km.key)}`
    );
    assert.deepStrictEqual(
      km.nonceBase,
      expectedNonceBase,
      `NonceBase mismatch.\nExpected: ${bytesToHex(expectedNonceBase)}\nGot: ${bytesToHex(km.nonceBase)}`
    );
  });
});
